File: leetcode/leetcode_cpp/dota2-senate.cpp
/*
https://leetcode.com/problems/dota2-senate

1. Queue
time: o(n)
space: o(n)
*/
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> r1;
        queue<int> d1;
        int n = senate.size();
        for (int i=0; i<senate.size(); ++i) {
            if (senate[i] == 'R') r1.push(i);
            else d1.push(i);
        }

        while (!r1.empty() && !d1.empty()) {
            if (r1.front() < d1.front()) r1.push(n);
            else d1.push(n);
            n += 1;
            r1.pop(); d1.pop();
        }
        return r1.empty() ? "Dire" : "Radiant";
    }
};

File: leetcode/leetcode_py/dota2-senate.py
'''
https://leetcode.com/problems/dota2-senate

1. Queue
time: o(n)
space: o(n)
'''
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r = []
        d = []
        n = len(senate)
        for i, c in enumerate(senate):
            if c == 'R': r.append(i)
            else: d.append(i)
        while r and d:
            if r[0] < d[0]: r.append(r[0] + n)
            else: d.append(d[0] + n)
            n += 1
            r.pop(0)
            d.pop(0)
        return "Radiant" if r else "Dire"

        